### ISSUE TITLE
GraphQL request to pending messages cannot return Message

### ISSUE THREAD

**Issue by kikakkz** [linera-io/linera-protocol#2694]
**Description**
I add a new methond in `linera-service/src/node_service.rs` to export pending messages to CheCko frontend. I think it worked before (need to confirm with v0.11.3) due to we use this method for CheCko demo. But at v0.12.1 and latest main branch it won't return Message.

**Steps**
1  Add the following method in `linera-service/src/node_service.rs`
```
/// Returns the pending message of the chain                                                                                            
async fn pending_messages(&self, chain_id: ChainId) -> Result<Vec<IncomingBundle>, Error> {                                                                                                                                                                                   
    let client = self.context.lock().await.make_chain_client(chain_id)?;                                                                
    Ok(client.pending_message_bundles().await?)                                                                                         
}
```
2  Compile then run `linera service --listener-skip-process-inbox` then node service won't process inbox messages
3  Deploy an application in which it'll create a new message in the operation routine
4  Request from graphiql with following payload
```
query {
  pendingMessages(chainId:"YOUR_CHAIN_ID") {
    bundle {
      messages {
        authenticatedSigner
        grant
        refundGrantTo
        message
        kind
        index
      }
    }
  }
}
```

**Expectation**
It should return graphql object correctly.

**Result**
It returned
```
{
  "data": {
    "pendingMessages": [
      {
        "bundle": {
          "messages": [
            {
              "authenticatedSigner": null,
              "grant": "0.",
              "refundGrantTo": null,
              "message": null,
              "kind": "Protected",
              "index": 0
            }
          ]
        }
      }
    ]
  }
}
```
where the pending messages is printed as following
```
IncomingBundle { origin: Origin { sender: e476187f6ddfeb9d, medium: Direct }, bundle: MessageBundle { height: BlockHeight(9), timestamp: Timestamp(1729655427455600), certificate_hash: edcde041d763c609, transaction_index: 0, messages: [PostedMessage { authenticated_signer: None, grant: Amount(0), refund_grant_to: None, kind: Protected, index: 0, message: System(OpenChain(OpenChainConfig { ownership: ChainOwnership { super_owners: {Owner(e93e64876e17f406): 5c5711f4ae18ec5b}, owners: {}, multi_leader_rounds: 2, timeout_config: TimeoutConfig { fast_round_duration: None, base_timeout: TimeDelta(10000000), timeout_increment: TimeDelta(1000000), fallback_duration: TimeDelta(86400000000) } }, admin_id: e476187f6ddfeb9d, epoch: Epoch(0), committees: {Epoch(0): Committee { validators: {ValidatorName(659eb8693e0f48df): ValidatorState { network_address: "grpc:127.0.0.1:9000", votes: 100 }}, total_votes: 100, quorum_threshold: 67, validity_threshold: 34, policy: ResourceControlPolicy { block: Amount(0), fuel_unit: Amount(0), read_operation: Amount(0), write_operation: Amount(0), byte_read: Amount(0), byte_written: Amount(0), byte_stored: Amount(0), operation: Amount(0), operation_byte: Amount(0), message: Amount(0), message_byte: Amount(0), maximum_fuel_per_block: 18446744073709551615, maximum_executed_block_size: 18446744073709551615, maximum_bytes_read_per_block: 18446744073709551615, maximum_bytes_written_per_block: 18446744073709551615 } }}, balance: Amount(100000000000000000000), application_permissions: ApplicationPermissions { execute_operations: None, mandatory_applications: [], close_chain: [] } })) }] }, action: Accept }]
```

**Version**
`v0.12.1` and `main`

**Comment by kikakkz** [comment]
@jvff We may need some special steps to reproduce this issue. I'm making one step by step reproducible example for that. Please wait for me to clarify it.

**Comment by kikakkz** [comment]
To help the analysis of the issue, I do some work to clarify the reproducible steps.

1 `git clone https://github.com/kikakkz/linera-protocol.git`, it contains the exported pending messages apis
2 `cargo run --bin linera-schema-export > linera-service-graphql-client/gql/service_schema.graphql`. Generate new schema with added api.
3 `cargo install --path linera-service --features storage-service`
4 `cargo install --path linera-storage-service --features storage-service`
5 Then we run `linera net up` to start a local testnet
6 `cd linera-protocol/examples/counter; cargo build --release --target wasm32-unknown-unknown`
7 `linera publish-bytecode ../target/wasm32-unknown-unknown/release/counter_{contract,service}.wasm`
8 Run `linera service --listener-skip-process-inbox`, query with
```
query {
  chains {
    default
    list
  }
  pendingMessages(chainId:"368717fa4077e85e9d66ed96d85dad25f06029dc3cc986deef8cb81e6fea3483") {
    bundle {
      messages {
        authenticatedSigner
        grant
        refundGrantTo
        message
      }
    }
  }
}
```
```
At that time, there will be a ApplicationCreated message in the new microchain's inbox. And it could be queried correctly.
```
![image](https://github.com/user-attachments/assets/bf7c2378-db11-4b9d-912d-2a739953b8aa)
9 Quit the service in step 8, then run a faucet with initial wallet of the local testnet with `linera faucet --port 6000 --amount 100`
10 Create a new wallet from faucet `linera wallet init --faucet http://localhost:6000 --with-new-chain`
11 Run new service without process inbox `linera service --port 30080 --listener-skip-process-inbox`, query with the same graphql schema as above
```
At this time there will be a OpenChain message in the new microchain's inbox. And it couldn't be queried. The message field is null in graphql result.
```
12 In the added method of node_service.rs, it print as following in which `message` field of the `bundle` is an `OpenChain`, not null.
```
Pending messages: [IncomingBundle { origin: Origin { sender: e476187f6ddfeb9d, medium: Direct }, bundle: MessageBundle { height: BlockHeight(9), timestamp: Timestamp(1729752342474834), certificate_hash: 11e36c38776c1315, transaction_index: 0, messages: [PostedMessage { authenticated_signer: None, grant: Amount(0), refund_grant_to: None, kind: Protected, index: 0, message: System(OpenChain(OpenChainConfig { ownership: ChainOwnership { super_owners: {Owner(8e454f361e32cbcb): 01a1710422581b80}, owners: {}, multi_leader_rounds: 2, timeout_config: TimeoutConfig { fast_round_duration: None, base_timeout: TimeDelta(10000000), timeout_increment: TimeDelta(1000000), fallback_duration: TimeDelta(86400000000) } }, admin_id: e476187f6ddfeb9d, epoch: Epoch(0), committees: {Epoch(0): Committee { validators: {ValidatorName(b2c064d4ba423756): ValidatorState { network_address: "grpc:localhost:9000", votes: 100 }}, total_votes: 100, quorum_threshold: 67, validity_threshold: 34, policy: ResourceControlPolicy { block: Amount(0), fuel_unit: Amount(0), read_operation: Amount(0), write_operation: Amount(0), byte_read: Amount(0), byte_written: Amount(0), byte_stored: Amount(0), operation: Amount(0), operation_byte: Amount(0), message: Amount(0), message_byte: Amount(0), maximum_fuel_per_block: 18446744073709551615, maximum_executed_block_size: 18446744073709551615, maximum_bytes_read_per_block: 18446744073709551615, maximum_bytes_written_per_block: 18446744073709551615 } }}, balance: Amount(100000000000000000000), application_permissions: ApplicationPermissions { execute_operations: None, mandatory_applications: [], close_chain: [] } })) }] }, action: Accept }]
```

**Comment by jvff** [comment]
Thanks for the detailed steps! Again they are really useful to nail down the issue. I had to tweak a few things to reproduce though, let me know if there are any issues with them:

- Step 2 doesn't seem to be needed (I think you already included the changes in commit, so thanks!)
- I had to add a step right after step 7 that's `linera create-application BYTECODE_ID_FROM_STEP_7 --json-argument 10`

From what I can tell, the problem is the way we convert `Message` to an `async_graphql::Value`, in the following line:

https://github.com/linera-io/linera-protocol/blob/a0003ce22363e883da88a641910c7acbbe097764/linera-execution/src/lib.rs#L1200-L1203

That macro calls `async_graphql`'s internal macro, which calls the `async_graphql_value::to_value` function:

https://github.com/async-graphql/async-graphql/blob/c572a411cd25bfd67fef54208c58d8e4b371e0a7/src/resolver_utils/scalar.rs#L147

Which serializes the value using a custom deserializer. The custom deserializer apparently does not support maps properly, as it seems to return an error if the key is not a string:

https://github.com/async-graphql/async-graphql/blob/c572a411cd25bfd67fef54208c58d8e4b371e0a7/value/src/serializer.rs#L438-L632

This leads the `to_value` method generated by `doc_scalar` to return `null`, because I believe we use quite a few `BTreeMap`s in types used by `SystemMessage::OpenChain` where the key is not a string:

https://github.com/linera-io/linera-protocol/blob/a0003ce22363e883da88a641910c7acbbe097764/linera-execution/src/system.rs#L103
https://github.com/linera-io/linera-protocol/blob/a0003ce22363e883da88a641910c7acbbe097764/linera-base/src/ownership.rs#L51-L53

One workaround is to use `linera_base::bcs_scalar` instead of `doc_scalar` for the `Message` type. I'll keep investigating to see if there's a better solution.

**Comment by jvff** [comment]
I believe #2722 implements a better solution than the proposed workaround. Would you be able to test it to see if it fixes the issue for your use case?

**Comment by kikakkz** [comment]
> Thanks for the detailed steps! Again they are really useful to nail down the issue. I had to tweak a few things to reproduce though, let me know if there are any issues with them:
> 
> * Step 2 doesn't seem to be needed (I think you already included the changes in commit, so thanks!)
> * I had to add a step right after step 7 that's `linera create-application BYTECODE_ID_FROM_STEP_7 --json-argument 10`

Sorry I miss this step, 😄. Yeah, we should create-application here.

**Comment by kikakkz** [comment]
> I believe #2722 implements a better solution than the proposed workaround. Would you be able to test it to see if it fixes the issue for your use case?

I'll test and feedback soon.

**Comment by kikakkz** [comment]
> Thanks for the detailed steps! Again they are really useful to nail down the issue. I had to tweak a few things to reproduce though, let me know if there are any issues with them:
> 
> * Step 2 doesn't seem to be needed (I think you already included the changes in commit, so thanks!)
> * I had to add a step right after step 7 that's `linera create-application BYTECODE_ID_FROM_STEP_7 --json-argument 10`
> 
> From what I can tell, the problem is the way we convert `Message` to an `async_graphql::Value`, in the following line:
> 
> https://github.com/linera-io/linera-protocol/blob/a0003ce22363e883da88a641910c7acbbe097764/linera-execution/src/lib.rs#L1200-L1203
> 
> That macro calls `async_graphql`'s internal macro, which calls the `async_graphql_value::to_value` function:
> 
> https://github.com/async-graphql/async-graphql/blob/c572a411cd25bfd67fef54208c58d8e4b371e0a7/src/resolver_utils/scalar.rs#L147
> 
> Which serializes the value using a custom deserializer. The custom deserializer apparently does not support maps properly, as it seems to return an error if the key is not a string:
> 
> https://github.com/async-graphql/async-graphql/blob/c572a411cd25bfd67fef54208c58d8e4b371e0a7/value/src/serializer.rs#L438-L632
> 
> This leads the `to_value` method generated by `doc_scalar` to return `null`, because I believe we use quite a few `BTreeMap`s in types used by `SystemMessage::OpenChain` where the key is not a string:
> 
> https://github.com/linera-io/linera-protocol/blob/a0003ce22363e883da88a641910c7acbbe097764/linera-execution/src/system.rs#L103
> 
> 
> https://github.com/linera-io/linera-protocol/blob/a0003ce22363e883da88a641910c7acbbe097764/linera-base/src/ownership.rs#L51-L53
> 
> One workaround is to use `linera_base::bcs_scalar` instead of `doc_scalar` for the `Message` type. I'll keep investigating to see if there's a better solution.

Thanks to the detail reply. It's really useful to understand async-graphql serializer and deserializer deeper.

**Comment by kikakkz** [comment]
> linera service --port 30080 --listener-skip-process-inbox

I confirm the issue is fixed with commit `4b7fb68b4bb734b012b3cb4b1372245dc232ca36` in https://github.com/linera-io/linera-protocol/pull/2722.

![image](https://github.com/user-attachments/assets/680da1d7-31f6-4856-a9fa-520fbee62103)
